mllib/src/main/scala/org/apache/spark/ml/tree/impl/RandomForest.scala

```diff
@@ -1110,4 +1124,40 @@ private[spark] object RandomForest extends Logging {
     }
   }
 
+  /**
+   * Class for queueing nodes to split on each iteration.
+   * This is a FILO queue.
+   */
+  private[impl] class NodeQueue {
+    private var q: mutable.MutableList[(Int, LearningNode)] =
+      mutable.MutableList.empty[(Int, LearningNode)]
+
+    /** Put (treeIndex, node) into queue. Constant time. */
+    def put(treeIndex: Int, node: LearningNode): Unit = {
+      q += ((treeIndex, node))
+    }
+
+    /** Remove and return last inserted element.  Linear time (unclear in Scala docs though) */
```

Comment:
Note: This is not ideal, but it should be insignificant compared to the cost of communicating the trees.  I am not aware of an existing solution for a stack with constant-time push/pop in Scala.